pkg/scheduler/framework/plugins/clusteraffinity/filtering_test.go

```diff
@@ -0,0 +1,343 @@
+						RequiredDuringSchedulingIgnoredDuringExecution: &fleetv1beta1.ClusterSelector{
+							ClusterSelectorTerms: []fleetv1beta1.ClusterSelectorTerm{
+								{
+									LabelSelector: metav1.LabelSelector{
+										MatchLabels: map[string]string{"region": "us-west"},
+									},
+								},
+							},
+						},
+						PreferredDuringSchedulingIgnoredDuringExecution: []fleetv1beta1.PreferredClusterSelector{
+							{
+								Weight: 5,
+								Preference: fleetv1beta1.ClusterSelectorTerm{
+									LabelSelector: metav1.LabelSelector{
+										MatchLabels: map[string]string{
+											"region": "us-west",
+										},
+									},
+								},
+							},
+							{
+								Weight: 1,
+								Preference: fleetv1beta1.ClusterSelectorTerm{
+									LabelSelector: metav1.LabelSelector{
+										MatchLabels: map[string]string{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			want: nil, // not skip the filter stage
+			wantPluginState: &pluginState{
+				requiredAffinityTerms: []affinityTerm{
+					{
+						selector: labels.SelectorFromSet(map[string]string{"region": "us-west"}),
+					},
+				},
+				preferredAffinityTerms: []preferredAffinityTerm{
+					{
+						weight: 5,
+						affinityTerm: affinityTerm{
+							selector: labels.SelectorFromSet(map[string]string{"region": "us-west"}),
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			state := framework.NewCycleState(nil)
+			snapshot := &fleetv1beta1.ClusterSchedulingPolicySnapshot{
+				Spec: fleetv1beta1.SchedulingPolicySnapshotSpec{
+					Policy: tc.policy,
+				},
+			}
+			p := New()
+			got := p.PreFilter(context.Background(), state, snapshot)
+			if diff := cmp.Diff(tc.want, got, cmpStatusOptions); diff != "" {
+				t.Errorf("PreFilter() status mismatch (-want, +got):\n%s", diff)
+			}
+			if tc.wantPluginState == nil {
+				return
+			}
+			val, err := state.Read(framework.StateKey(p.Name()))
+			if err != nil {
+				t.Fatalf("Read() state got %v, want not nil", err)
+			}
+			gotPluginState, ok := val.(*pluginState)
+			if !ok {
+				t.Fatalf("failed to cast value %v to the pluginState type", val)
+			}
+			if diff := cmp.Diff(tc.wantPluginState, gotPluginState, cmpPluginStateOptions); diff != "" {
+				t.Errorf("Read() pluginState mismatch (-want, +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestFilter(t *testing.T) {
+	tests := []struct {
+		name              string
+		ps                *pluginState
+		notSetPluginState bool
+		cluster           *fleetv1beta1.MemberCluster
+		want              *framework.Status
+	}{
+		{
+			name:              "pluginState is not set",
+			notSetPluginState: true,
+			want:              framework.FromError(errors.New("invalid state"), defaultClusterAffinityPluginOptions.name),
+		},
+		{
+			name: "nil pluginState",
+			want: framework.FromError(errors.New("invalid state"), defaultClusterAffinityPluginOptions.name),
```

Comment:
nit: defaultClusterAffinityPluginOptions.name is better made into a top-level constant.